src/dstore/pmix_esh.c

```diff
@@ -0,0 +1,1374 @@
+{
+    ns_track_elem_t *new_elem = NULL;
+
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s: nspace %s", __FILE__, __LINE__, __func__, nspace));
+
+    /* check if this namespace is already being tracked to avoid duplicating data. */
+    PMIX_LIST_FOREACH(new_elem, &namespace_info_list, ns_track_elem_t) {
+        if (0 == strncmp(nspace, new_elem->ns_name, PMIX_MAX_NSLEN+1)) {
+            /* data for this namespace should be already stored in shared memory region. */
+            /* so go and just put new data. */
+            PMIX_OUTPUT_VERBOSE((7, pmix_globals.debug_output,
+                        "%s:%d:%s: found nspace %s in the track list", __FILE__, __LINE__, __func__, nspace));
+            return new_elem;
+        }
+    }
+
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s: create new object for nspace %s", __FILE__, __LINE__, __func__, nspace));
+    /* create shared memory regions for this namespace and store its info locally
+     * to operate with address and detach/unlink afterwards. */
+    new_elem = PMIX_NEW(ns_track_elem_t);
+    strncpy(new_elem->ns_name, nspace, PMIX_MAX_NSLEN+1);
+
+    pmix_list_append(&namespace_info_list, &new_elem->super);
+
+    return new_elem;
+}
+
+static rank_meta_info *_get_rank_meta_info(int rank, seg_desc_t *segdesc)
+{
+    PMIX_OUTPUT_VERBOSE((10, pmix_globals.debug_output,
+                         "%s:%d:%s", __FILE__, __LINE__, __func__));
+    size_t i;
+    rank_meta_info *elem = NULL;
+    seg_desc_t *tmp = segdesc;
+    size_t num_elems, rel_offset;
+    int id;
+    rank_meta_info *cur_elem;
+    if (1 == direct_mode) {
+        /* do linear search to find the requested rank inside all meta segments
+         * for this namespace. */
+        /* go through all existing meta segments for this namespace */
+        do {
+            num_elems = *((size_t*)(tmp->seg_info.seg_base_addr));
+            for (i = 0; i < num_elems; i++) {
+                cur_elem = (rank_meta_info*)((uint8_t*)(tmp->seg_info.seg_base_addr) + sizeof(size_t) + i * sizeof(rank_meta_info));
+                if (rank == (int)cur_elem->rank) {
+                    elem = cur_elem;
+                    break;
+                }
+            }
+            tmp = tmp->next;
+        }
+        while (NULL != tmp && NULL == elem);
+    } else {
+        /* directly compute index of meta segment (id) and relative offset (rel_offset)
+         * inside this segment for fast lookup a rank_meta_info object for the requested rank. */
+        id = rank/max_meta_elems;
+        rel_offset = (rank%max_meta_elems) * sizeof(rank_meta_info) + sizeof(size_t);
+        /* go through all existing meta segments for this namespace.
+         * Stop at id number if it exists. */
+        while (NULL != tmp->next && 0 != id) {
+            tmp = tmp->next;
+            id--;
+        }
+        if (0 == id) {
```

Comment:
Maybe we need to attach new segments that was created since we access this NS last time?
Is it guaranteed here that all meta-segments was mapped? If so, where is it?